MTA-7135: Agent Mode Compilation Guardrails AST Verification (#380)
migtools/mta-documentation: docs/topics/developer-lightspeed/con_agent-mode-guardrails.adoc

// Module included in the following assemblies:
//
// * docs/topics/developer-lightspeed/assembly_understanding_dl.adoc

:_mod-docs-content-type: CONCEPT

[id="agent-mode-guardrails_{context}"]
= Output verification and agent mode guardrails

[role="_abstract"]

> In agent mode, the {mta-dl-plugin} includes an output verification mechanism that acts as a native compiler validation loop. This mechanism ensures that AI-generated code suggestions are syntactically valid and compatible with your classpath before they are written to disk.

Reviewer: WDYT?
```suggestion
To understand how the system ensures syntactically valid code, review the output verification mechanism. In agent mode, the {mta-dl-plugin} acts as a native compiler validation loop.

The verification mechanism ensures that AI-generated code suggestions are syntactically valid and compatible with your classpath before they are written to disk.
```

The verification mechanism provides a programmatic guardrail against common large language model (LLM) risks:

* Hallucinated application programming interface (API) signatures
* References to removed or deprecated methods
* Syntactically broken output

When you run an analysis in agent mode, the {mta-dl-plugin} passes each code suggestion through a local build dry run before applying it to your workspace. This process runs entirely on your developer workstation. No source code or build artifacts leave your local environment during verification.

The extension invokes the standard build toolchain that is already configured for your project directly in your local workspace, such as the following toolchains:

`Java projects`:: Apache Maven (`mvn compile`) or Gradle (`gradle compileJava`).

> `JavaScript and TypeScript projects`:: npm (`npm run build`) or Yarn (`yarn build`).

Reviewer: Let's turn it into a simple list. There are no big explanations, so I think it's enough to just have a bulleted list instead of a definition one. WDYT?
```suggestion
* `Java projects`: Apache Maven (`mvn compile`) or Gradle (`gradle compileJava`).
* `JavaScript and TypeScript projects`: npm (`npm run build`) or Yarn (`yarn build`).
```

// Module included in the following assemblies:
//
// * docs/topics/developer-lightspeed/assembly_understanding_dl.adoc

:_mod-docs-content-type: CONCEPT
[id="agent-mode-guardrails_{context}"]
= Output verification and agent mode guardrails

[role="_abstract"]
In agent mode, {mta-dl-plugin} includes an output verification mechanism that acts as a native compiler validation loop. This mechanism ensures that AI-generated code suggestions are syntactically valid and compatible with your classpath before they are written to disk.

> This verification mechanism provides a programmatic guardrail against common large language model (LLM) risks, such as hallucinated application programming interface (API) signatures, references to removed or deprecated methods, and syntactically broken output.

Reviewer: Just because folks ask us to have the short description and the following paragraph independent + suggested to have a list for better readability. WDYT?
```suggestion
The verification mechanism provides a programmatic guardrail against common large language model (LLM) risks, for example:

 * Hallucinated application programming interface (API) signatures
 * References to removed or deprecated methods
 * Syntactically broken output
```

== Agent mode code quality safeguards

When you run an analysis in agent mode, {mta-dl-plugin} passes each code suggestion through a local build dry run before applying it to your workspace. This process runs entirely on your developer workstation. No source code or build artifacts leave your local environment during verification.

The extension invokes the standard build toolchain that is already configured for your project directly in your local workspace:

* *Java projects*: Apache Maven (`mvn compile`) or Gradle (`gradle compileJava`)
* *JavaScript and TypeScript projects*: npm (`npm run build`) or Yarn (`yarn build`)

The dry run executes directly on the underlying file system of your local workspace without containerization or process isolation. If the dry run exits cleanly, {mta-dl-plugin} presents the verified suggestion for your approval before applying it to the workspace. If the dry run reports errors, agent mode attempts automatic correction.

== Agent mode validation checks

Agent mode performs the following verification checks on every code suggestion:

> Syntax verification::

Reviewer: Just a thought: Since you used a standard bulleted list in the previous subsection, do you think it might be better to use it here as well for consistency?

// Module included in the following assemblies:
//
// * docs/topics/developer-lightspeed/assembly_understanding_dl.adoc

:_mod-docs-content-type: CONCEPT
[id="agent-mode-guardrails_{context}"]
= Output verification and agent mode guardrails

[role="_abstract"]
In agent mode, {mta-dl-plugin} includes an output verification mechanism that acts as a native compiler validation loop. This mechanism ensures that AI-generated code suggestions are syntactically valid and compatible with your classpath before they are written to disk.

> This verification mechanism provides a programmatic guardrail against common large language model (LLM) risks, such as hallucinated API signatures, references to removed or deprecated methods, and syntactically broken output.

Reviewer 2: ```suggestion
This verification mechanism provides a programmatic guardrail against common large language model (LLM) risks, such as hallucinated application programming interface (API) signatures, references to removed or deprecated methods, and syntactically broken output.
```

== How agent mode safeguards code quality

When you run an analysis in agent mode, {mta-dl-plugin} passes each code suggestion through a local build dry-run before applying it to your workspace. This process runs entirely on your developer workstation; no source code or build artifacts leave your local environment during verification.

> The extension invokes the standard build toolchain already configured for your project within an isolated local pipeline:

Author: ```suggestion
The extension invokes the standard build toolchain already configured for your project directly in your local workspace:
```

> * *Java projects*: Apache Maven (`mvn compile`) or Gradle (`gradle compileJava`)

Reviewer 2: Please could you confirm if Developer Lightspeed compiles a project without pom/build files too?

* *JavaScript and TypeScript projects*: npm (`npm run build`) or Yarn (`yarn build`)

> The dry-run executes in an isolated local container separate from your project's normal build lifecycle, so verification does not trigger deployment tasks or modify build caches. If the dry-run exits cleanly, {mta-dl-plugin} applies the suggestion to the workspace. If the dry-run reports errors, agent mode attempts automatic correction, as described in the following section.

Author: ```suggestion
The dry-run executes directly on the underlying filesystem of your local workspace without containerization or process isolation. If the dry-run exits cleanly, {mta-dl-plugin} presents the verified suggestion for your approval before applying it to the workspace. If the dry-run reports errors, agent mode attempts automatic correction, as described in the following section.
```

Import and dependency consistency::
Agent mode confirms that every `import` statement that the suggestion adds corresponds to a type that is available on the class path. Agent mode flags missing or incorrect imports for correction.

== Autocorrection on compilation failure

When a compilation step fails, agent mode does not immediately surface the error to you. Instead, {mta-dl-plugin} feeds the compiler diagnostic output back into the {mta-dl-full} model as additional context and requests a corrected suggestion. This retry loop runs up to two times for minor errors, such as a single unresolved import or a mismatched method argument type.

If the suggestion cannot be corrected within the retry limit, agent mode surfaces the original compiler output alongside the suggestion and marks the code as unverified. You can then review and manually resolve the issue before you apply the change.

[NOTE]
====

> Agent mode autocorrection is designed for minor, isolated errors. Address large-scale structural issues, such as a suggestion that fundamentally misunderstands your project's architecture, by refining your prompt rather than relying on autocorrection.

Reviewer 2: "by refining your prompt rather than relying on autocorrection." 

Can you please mention when the user gives a prompt for the dry run? IIUC, the prompt is auto-generated by Developer Lightspeed for the LLM but please cross verify the behavior.

Reviewer 3: Yeah, I would agree with this statement. The user can't enter any kind of manual prompt, everything is handled by our extension.

> ====

Author: ```suggestion
```
